clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,263 @@
+     * are no longer in the target set.
+     *
+     * @param assignment Holds the {@link Assignment} which includes the target set of topics
+     * @return Set of partitions to revoke
+     */
+
+    SortedSet<TopicPartition> getPartitionsToRevoke(Assignment assignment) {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
+        partitions.addAll(subscriptions.assignedPartitions());
+        partitions.removeAll(targetPartitions(assignment));
+        return partitions;
+    }
+
+    /**
+     * Determine which partitions are newly assigned. This is done by comparing the
+     * {@link Assignment#topicPartitions() target set from the assignment} against the
+     * {@link SubscriptionState#assignedPartitions() current set}. Any {@link TopicPartition partitions} from the
+     * target set that are not already in the current set are included in the returned set.
+     *
+     * @param assignment {@link Optional} that holds the {@link Assignment} which includes the target set of topics
+     * @return Set of partitions to assign
+     */
+
+    SortedSet<TopicPartition> getPartitionsToAssign(Assignment assignment) {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
+        partitions.addAll(targetPartitions(assignment));
+        partitions.removeAll(subscriptions.assignedPartitions());
+        return partitions;
+    }
+
+    /**
+     * Determine which partitions should be "lost". This is simply the
+     * {@link SubscriptionState#assignedPartitions() current set} of {@link TopicPartition partitions}.
+     *
+     * @return Set of partitions to "lose"
+     */
+    SortedSet<TopicPartition> getPartitionsToLose() {
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
+        partitions.addAll(subscriptions.assignedPartitions());
+        return partitions;
+    }
+
+    private SortedSet<TopicPartition> targetPartitions(Assignment assignment) {
+        Map<Uuid, String> topicIdToNameMap = createTopicIdToNameMapping();
+        SortedSet<TopicPartition> partitions = new TreeSet<>(new TopicPartitionComparator());
+
+        for (TopicPartitions topicPartitions : assignment.topicPartitions()) {
+            Uuid topicId = topicPartitions.topicId();
+            String topicName = Objects.requireNonNull(topicIdToNameMap.get(topicId), () -> String.format("No topic name was found in the metadata for topic ID %s", topicId));
+
+            for (Integer partition : topicPartitions.partitions()) {
+                partitions.add(new TopicPartition(topicName, partition));
+            }
+        }
+
+        return partitions;
+    }
+
+    private Map<Uuid, String> createTopicIdToNameMapping() {
+        Map<Uuid, String> map = new HashMap<>();
+
+        for (Map.Entry<String, Uuid> entry : metadata.topicIds().entrySet())
+            map.put(entry.getValue(), entry.getKey());
+
+        return map;
+    }
```

Comment:
That seems like a good idea. Perhaps as a separate PR?

Reply:
@dajac I added the changes for the topic mapping to this PR _for now_. I will split it out, but wanted to leave it here in case we end up going a different direction that obviates the need for this.